File: py_orchestrator/src/py_orchestrator/agent_log.py
from __future__ import annotations

from dataclasses import dataclass, field, asdict
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional


SCHEMA_VERSION = "agentic-ml-log.v1"


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass
class AgentIds:
    run_id: Optional[str] = None
    task_id: Optional[str] = None
    trace_id: Optional[str] = None
    span_id: Optional[str] = None
    parent_span_id: Optional[str] = None


@dataclass
class AgentSource:
    service: str = "py_orchestrator"
    component: str = "unknown"
    function: Optional[str] = None
    env: str = "dev"


@dataclass
class AgentContext:
    intent: Optional[str] = None
    segment: Optional[str] = None
    layer: Optional[str] = None
    step: Optional[str] = None


@dataclass
class ObservationEvidence:
    latency_ms: Optional[float] = None
    timeout_ms: Optional[float] = None
    http_status: Optional[int] = None
    extra: Dict[str, Any] = field(default_factory=dict)


@dataclass
class AgentObservation:
    summary: str
    raw_message: Optional[str] = None
    evidence: ObservationEvidence = field(default_factory=ObservationEvidence)


@dataclass
class AgentClassification:
    error_code: Optional[str] = None
    severity: Optional[str] = None  # e.g. "recoverable" / "fatal" / "info"
    retryable: Optional[bool] = None
    confidence: Optional[float] = None  # for LLM-assigned classifications


@dataclass
class AgentAction:
    taken: Optional[str] = None
    next_suggested: Optional[str] = None


@dataclass
class AgentLogEvent:
    """
    LLM-friendly observation event, designed to be emitted as NDJSON.

    This structure mirrors how we want agents to see logs and metrics:
    each line answers:
    - when did it happen?
    - for which run/task?
    - from which service/component?
    - at which step?
    - what happened?
    - what evidence do we have?
    - what action was taken or suggested?
    """

    schema_version: str = SCHEMA_VERSION
    ts: str = field(default_factory=utc_now_iso)
    level: str = "INFO"
    event: str = "generic.event"

    ids: AgentIds = field(default_factory=AgentIds)
    source: AgentSource = field(default_factory=AgentSource)
    context: AgentContext = field(default_factory=AgentContext)
    observation: AgentObservation = field(
        default_factory=lambda: AgentObservation(summary="no-summary")
    )
    classification: AgentClassification = field(
        default_factory=AgentClassification
    )
    action: AgentAction = field(default_factory=AgentAction)

    def to_dict(self) -> Dict[str, Any]:
        return asdict(self)


def to_ndjson_lines(events: List[AgentLogEvent]) -> str:
    """
    Convert a list of AgentLogEvent objects into an NDJSON string.

    Each line is a JSON object.
    """
    import json

    lines: List[str] = []
    for event in events:
        lines.append(json.dumps(event.to_dict(), separators=(",", ":")))
    return "\n".join(lines)

File: gateway/src/agent_log.rs
use serde::Serialize;
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Debug, Serialize)]
pub struct AgentIds {
    pub run_id: Option<String>,
    pub task_id: Option<String>,
    pub trace_id: Option<String>,
    pub span_id: Option<String>,
    pub parent_span_id: Option<String>,
}

impl Default for AgentIds {
    fn default() -> Self {
        Self {
            run_id: None,
            task_id: None,
            trace_id: None,
            span_id: None,
            parent_span_id: None,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct AgentSource {
    pub service: String,
    pub component: String,
    pub function: Option<String>,
    pub env: String,
}

impl Default for AgentSource {
    fn default() -> Self {
        Self {
            service: "gateway".to_string(),
            component: "unknown".to_string(),
            function: None,
            env: "dev".to_string(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct AgentContext {
    pub intent: Option<String>,
    pub segment: Option<String>,
    pub layer: Option<String>,
    pub step: Option<String>,
}

impl Default for AgentContext {
    fn default() -> Self {
        Self {
            intent: None,
            segment: None,
            layer: None,
            step: None,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct ObservationEvidence {
    pub latency_ms: Option<f64>,
    pub timeout_ms: Option<f64>,
    pub http_status: Option<i32>,
    #[serde(default)]
    pub extra: BTreeMap<String, Value>,
}

impl Default for ObservationEvidence {
    fn default() -> Self {
        Self {
            latency_ms: None,
            timeout_ms: None,
            http_status: None,
            extra: BTreeMap::new(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct AgentObservation {
    pub summary: String,
    pub raw_message: Option<String>,
    pub evidence: ObservationEvidence,
}

impl Default for AgentObservation {
    fn default() -> Self {
        Self {
            summary: "no-summary".to_string(),
            raw_message: None,
            evidence: ObservationEvidence::default(),
        }
    }
}

#[derive(Debug, Serialize)]
pub struct AgentClassification {
    pub error_code: Option<String>,
    pub severity: Option<String>,
    pub retryable: Option<bool>,
    pub confidence: Option<f64>,
}

impl Default for AgentClassification {
    fn default() -> Self {
        Self {
            error_code: None,
            severity: None,
            retryable: None,
            confidence: None,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct AgentAction {
    pub taken: Option<String>,
    pub next_suggested: Option<String>,
}

impl Default for AgentAction {
    fn default() -> Self {
        Self {
            taken: None,
            next_suggested: None,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct AgentLogEvent {
    pub schema_version: String,
    pub ts: String,
    pub level: String,
    pub event: String,
    pub ids: AgentIds,
    pub source: AgentSource,
    pub context: AgentContext,
    pub observation: AgentObservation,
    pub classification: AgentClassification,
    pub action: AgentAction,
}

impl Default for AgentLogEvent {
    fn default() -> Self {
        Self {
            schema_version: "agentic-ml-log.v1".to_string(),
            ts: chrono::Utc::now().to_rfc3339(),
            level: "INFO".to_string(),
            event: "generic.event".to_string(),
            ids: AgentIds::default(),
            source: AgentSource::default(),
            context: AgentContext::default(),
            observation: AgentObservation::default(),
            classification: AgentClassification::default(),
            action: AgentAction::default(),
        }
    }
}